library/core/src/marker.rs

```diff
@@ -214,6 +268,20 @@ pub trait StructuralEq {
     // Empty.
 }
 
+// FIXME: Remove special cases of these types from the compiler pattern checking code and always check `T: StructuralEq` instead
+marker_impls! {
+    #[unstable(feature = "structural_match", issue = "31434")]
+    StructuralEq for
+        usize, u8, u16, u32, u64, u128,
+        isize, i8, i16, i32, i64, i128,
+        bool,
+        char,
+        str /* Technically requires `[u8]: StructuralEq` */,
+        {T: StructuralEq, const N: usize} [T; N],
+        {T: StructuralEq} [T],
+        {T: ?Sized + StructuralEq} &T,
```

Comment:
```suggestion
        str,
        {T, const N: usize} [T; N],
        {T} [T],
        {T: ?Sized} &T,
```
`StructuralEq` is shallow (i.e. doesn't care if the fields also implement `StructuralEq` or not) so these dont need any bounds as their equality conceptually just defers to its fields

Reply:
This is [true](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=bd3e01af7455fe9d68fbbcbf5d65771c), but doesn't make any sense to me...

The match is still [prohibited](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=63aff8fa7be36ebae2247a32d3dcb72d) even if the type itself is `StructuralEq`, but its fields are not... Why do we even implement/have `StructuralEq` if its shallow? Or more like... wouldn't it be better, if it wasn't shallow and `T: StructuralEq` would allow to match?...

Reply:
I don't know why its structured like that for pattern matching but this is how it works :sweat_smile:

Reply:
Okay, I'll try fixing it later then (_dies of too long todo-list_)

Reply:
this seems to have not been fixed

Reply:
Ooops. Fixed.